### PR TITLE
Add gradient functions

### PR DESCRIPTION
**Pull request by gmuraru**
## Description
Add the possibility to train a neural network:
* [x] added the forward pass for `mul` and `t`
* [x] Add the backwards pass
* [ ] More functions should be added: TBD in some subsequent PRs
* [ ] Add tests: TBD in **this** PR
   *  [x] Grad funcs 
   *  [ ] MPC Tensor backward

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: src/sympc/tensor/test2.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import torch
```
Comment (gmuraru) [inline src/sympc/tensor/test2.py:1]: Those files are only for testing
TODO: More test cases should be created.

File: src/sympc/tensor/share_tensor.py [hunk 2]
```diff
@@ -95,11 +118,16 @@ def __init__(
...
 
         self.tensor: Optional[torch.Tensor] = None
-
         if data is not None:
             tensor_type = self.session.tensor_type
             self.tensor = self._encode(data).type(tensor_type)
 
+        self.grad_fn = None
+        self.grad = 0
+        self.ctx = {}
```
Comment (gmuraru) [inline src/sympc/tensor/share_tensor.py:127]: TODO: Change this to ```cache_training_ctx```

File: test.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+from jax import grad
```
Comment (gmuraru) [inline test.py:1]: TODO: Remove this

File: src/sympc/tensor/share_tensor.py [hunk 4]
```diff
@@ -288,18 +320,53 @@ def div(self, y: Union[int, float, torch.Tensor, "ShareTensor"]) -> "ShareTensor
...
+            return
+
+        if self.grad_fn is None:
+            raise ValueError(f"Do not know how to propagate {self}")
+
+        grad = self.grad_fn.backward(self.ctx, self.grad)
+        if not isinstance(grad, (list, tuple)):
+            grad = (grad,)
+
+        for idx, parent in enumerate(self.parents):
```
Comment (gmuraru) [inline src/sympc/tensor/share_tensor.py:365]: TODO: Use zip

File: src/sympc/tensor/share_tensor.py [hunk 5]
```diff
@@ -288,18 +320,53 @@ def div(self, y: Union[int, float, torch.Tensor, "ShareTensor"]) -> "ShareTensor
...
+            print(parent)
+            parent.backward(grad_input=grad[idx])
+
```
Comment (gmuraru) [inline src/sympc/tensor/share_tensor.py:368]: TODO: Clear the context and gradients at the end.
If leaf_node we should leave the gradient

File: tests/sympc/tensor/grad/grad_functions_test.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+    assert ctx["x_shape"] == (2, 3)
+
+    res = res_mpc.reconstruct()
+    expected = x.sum()
+
+    assert (res == expected).all()
+
+
+def test_grad_sum_bacward(get_clients) -> None:
```
Comment (aanurraj) [inline tests/sympc/tensor/grad/grad_functions_test.py:111]: TYPO: backward

File: src/sympc/tensor/mpc_tensor.py [hunk 7]
```diff
@@ -695,6 +754,80 @@ def __repr__(self):
...
+        """Get the attribute and check if we should track the gradient.
+
+        Args:
+            attr_name (str): The name of the attribute
+
+        Returns:
+            The attribute specific for this instance
+        """
+        # TODO: Fix this
+        from sympc.tensor.grads import GRAD_FUNCS
```
Comment (aanurraj) [inline src/sympc/tensor/mpc_tensor.py:767]: Q: why we are importing this in middle ? do we get any error ?

**Review by aanurraj — APPROVED** [review]
Looks Good 🎉

File: src/sympc/tensor/mpc_tensor.py (hunk 7 above)
Comment (gmuraru) [reply src/sympc/tensor/mpc_tensor.py:767]: because of circular dependency - it has a TODO
